src/main/scala/zio/interop/javaconcurrent.scala

```diff
@@ -16,94 +16,152 @@
 
 package zio.interop
 
+import java.nio.channels.CompletionHandler
 import java.util.concurrent.{ CompletableFuture, CompletionException, CompletionStage, Future }
 
 import zio._
 import zio.blocking.{ blocking, Blocking }
 
 import scala.concurrent.ExecutionException
+import scala.util.control.NonFatal
 
 object javaconcurrent {
 
-  implicit class IOObjJavaconcurrentOps(private val taskObj: Task.type) extends AnyVal {
+  def withCompletionHandler[T](op: CompletionHandler[T, Unit] => Unit): Task[T] =
+    Task.effectAsync[T] { k =>
+      val handler = new CompletionHandler[T, Unit] {
+        def completed(result: T, u: Unit): Unit =
+          k(Task.succeed(result))
 
-    private def unsafeCompletionStageToIO[A](cs: CompletionStage[A]): Task[A] =
-      IO.effectAsync { cb =>
+        def failed(t: Throwable, u: Unit): Unit =
+          t match {
+            case NonFatal(e) => k(Task.fail(e))
+            case _           => k(Task.die(t))
+          }
+      }
+
+      try {
+        op(handler)
+      } catch {
+        case NonFatal(e) => k(Task.fail(e))
```

Comment:
@pshirshov how do I get a `Platform` here, so that I could use its `fatal`?

Reply:
@sideeffffect You can get it beforehand with `Task.effectSuspendWith`, as in
```scala
Task.effectSuspendWith { platform =>
  Task.effectAsync[T] { k =>
     ...
  }
}
```

Reply:
thanks! applied. wasn't aware of those :)
is it OK now?